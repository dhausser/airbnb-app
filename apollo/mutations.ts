import { gql } from '@apollo/client'

export const CREATE_DRAFT_MUTATION = gql`
  mutation CreateDraft($title: String!, $content: String!, $authorEmail: String!) {
    createDraft(title: $title, content: $content, authorEmail: $authorEmail) {
      id
      title
      content
      author {
        email
      }
    }
  }
`

export const UPDATE_POST_MUTATION = gql`
  mutation UpdatePost($id: ID!, $title: String!, $content: String) {
    updateDraft(id: $id, title: $title, content: $content) {
      id
      title
      content
      published
    }
  }
`

export const DELETE_POST_MUTATION = gql`
  mutation DeletePost($id: ID!) {
    deletePost(id: $id) {
      id
      title
      content
      author {
        id
        email
        name
      }
      published
    }
  }
`

export const DELETE_POSTS_MUTATION = gql`
  mutation DeletePosts {
    deletePosts
  }
`
