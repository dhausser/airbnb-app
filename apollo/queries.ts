import { gql } from '@apollo/client'

export const GET_POSTS_QUERY = gql`
  query PostsQuery {
    posts {
      id
      title
      content
      author {
        email
      }
    }
  }
`

export const GET_POST_QUERY = gql`
  query PostQuery($id: ID!) {
    post(id: $id) {
      id
      title
      content
      author {
        email
      }
    }
  }
`
