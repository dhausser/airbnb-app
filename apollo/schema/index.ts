export * from './mutation'
export * from './post'
export * from './profile'
export * from './query'
export * from './user'
export * from './viewer'
