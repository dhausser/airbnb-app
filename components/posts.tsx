import { ApolloError } from '@apollo/client'
import { PostCard } from './post-card'
import { PostsQuery } from '../__generated__/PostsQuery'

interface Props {
  loading: boolean
  error: ApolloError
  data: PostsQuery
}

export const Posts: React.FC<Props> = ({ loading, error, data }) => {
  return (
    <>
      {loading ? (
        <div className="card">
          <p>Loading...</p>
        </div>
      ) : error ? (
        <div className="card">
          <p>{`${error.name}: ${error.message}`}</p>
        </div>
      ) : (
        data.posts.map((post) => <PostCard key={post.id} post={post} />)
      )}
    </>
  )
}
