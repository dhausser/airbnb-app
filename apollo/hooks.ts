import { useQuery, useMutation, ApolloError } from '@apollo/client'
import { GET_POSTS_QUERY, GET_POST_QUERY } from './queries'
import { DELETE_POSTS_MUTATION } from './mutations'

import * as PostQueryTypes from '../__generated__/PostQuery'
import * as PostsQueryTypes from '../__generated__/PostsQuery'
import * as DeletePostsMutationTypes from '../__generated__/DeletePosts'

interface QueryResults {
  loading: boolean
  error: ApolloError | undefined
  data: any
}

export const usePosts = (): QueryResults => {
  const { loading, error, data } = useQuery<PostsQueryTypes.PostsQuery>(GET_POSTS_QUERY)

  return {
    loading,
    error,
    data,
  }
}

export const usePost = (id: string | string[]): QueryResults => {
  const { loading, error, data } = useQuery<PostQueryTypes.PostQuery, PostQueryTypes.PostQueryVariables>(
    GET_POST_QUERY,
    {
      variables: { id: id as string },
    }
  )

  return {
    loading,
    error,
    data,
  }
}

export const useDeletePosts = (): any => {
  return useMutation<DeletePostsMutationTypes.DeletePosts>(DELETE_POSTS_MUTATION)
}
