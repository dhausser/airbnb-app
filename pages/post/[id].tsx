import { useRouter } from 'next/router'
import { GetServerSideProps } from 'next'
import { ApolloClient, NormalizedCacheObject } from '@apollo/client'
import { initializeApollo } from '../../apollo/client'

import { Post } from '../../components/post'
import { GET_POST_QUERY } from '../../apollo/queries'

function SinglePost() {
  const router = useRouter()
  const { id } = router.query

  return <Post id={id} />
}

export const getServerSideProps: GetServerSideProps = async (context) => {
  const apolloClient: ApolloClient<NormalizedCacheObject | null> = initializeApollo()

  await apolloClient.query({
    query: GET_POST_QUERY,
    variables: {
      id: context.params?.id,
    },
  })

  // Pass post data to the page via props
  return {
    props: {
      initialApolloState: apolloClient.cache.extract(),
    },
  }
}

export default SinglePost
