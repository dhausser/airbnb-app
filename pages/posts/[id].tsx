import { GetServerSideProps } from 'next'
import { useRouter } from 'next/router'
import { ApolloClient, NormalizedCacheObject } from '@apollo/client'

import { initializeApollo } from '../../apollo/client'
import { GET_POST_QUERY } from '../../apollo/queries'
import { usePost } from '../../apollo/hooks'
import { PostForm } from '../../components/post-form'
import { UpdateDraftButton, DeleteDraftButton } from '../../components/action-button'

import styles from '../../styles/Home.module.css'

export const SinglePost: React.FC = () => {
  const router = useRouter()
  const { id } = router.query
  const { loading, error, data } = usePost(id)

  if (error) return <p>{`${error.name}: ${error.message}`}</p>
  if (loading || !data) return <p>Loading...</p>

  const {
    title,
    content,
    author: { email: authorEmail },
  } = data.post

  return (
    <div className={styles.container}>
      <div className={styles.main}>
        <div className={styles.grid}>
          <PostForm initial={{ title, content: content as string, authorEmail }} />
          <UpdateDraftButton
            id={data.post.id}
            title={data.post.title}
            content={data.post.content}
            authorEmail={data.post.author.email}
          />
          <DeleteDraftButton id={id as string} />
        </div>
      </div>
    </div>
  )
}

export const getServerSideProps: GetServerSideProps = async (context) => {
  const apolloClient: ApolloClient<NormalizedCacheObject | null> = initializeApollo()

  await apolloClient.query({
    query: GET_POST_QUERY,
    variables: {
      id: context.params?.id,
    },
  })

  // Pass post data to the page via props
  return {
    props: {
      initialApolloState: apolloClient.cache.extract(),
    },
  }
}

export default SinglePost
