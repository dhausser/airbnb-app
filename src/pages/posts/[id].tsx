import { gql, ApolloClient } from '@apollo/client'
import { GetStaticPaths, GetStaticProps } from 'next'
import { initializeApollo } from '../../apollo/client'
import { PostsQuery } from '../index'
// import * as PostQueryTypes from './__generated__/PostQuery'

const PostQuery = gql`
  query PostQuery($id: ID!) {
    post(id: $id) {
      id
      title
      content
      author {
        email
      }
    }
  }
`

export const Post = ({ post }): JSX.Element => {
  // const router = useRouter()
  // const { id } = router.query

  // const { loading, error, data } = useQuery<
  //   PostQueryTypes.PostQuery,
  //   PostQueryTypes.PostQueryVariables
  // >(PostQuery, {
  //   variables: { id: id as string },
  // })

  // if (loading) return <p>Loading...</p>
  // if (error) return <p>{`${error.name}: ${error.message}`}</p>

  return (
    <div className="grid">
      <div className="card">
        <h3>{post.title} &rarr;</h3>
        <p>{post.content}</p>
        <p>{post.author.email}</p>
      </div>
    </div>
  )
}

// // This function gets called at build time
export const getStaticPaths: GetStaticPaths = async () => {
  const apolloClient = initializeApollo()

  // Call an external API endpoint to get posts
  const {
    data: { posts },
  } = await apolloClient.query({
    query: PostsQuery,
  })

  // Get the paths we want to pre-render based on posts
  const paths = posts.map((post) => `/posts/${post.id}`)

  // We'll pre-render only these paths at build time.
  // { fallback: false } means other routes should 404.
  return { paths, fallback: false }
}

// This also gets called at build time
export const getStaticProps: GetStaticProps = async ({ params }) => {
  const apolloClient: ApolloClient<{}> = initializeApollo()

  const {
    data: { post },
  } = await apolloClient.query({
    query: PostQuery,
    variables: {
      id: params.id,
    },
  })

  // Pass post data to the page via props
  return { props: { post } }
}

export default Post
