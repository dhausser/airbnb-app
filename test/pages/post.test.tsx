import { InMemoryCache } from '@apollo/client'
import { render, cleanup, waitFor } from '../test-utils'
import { SinglePost } from '../../pages/posts/[id]'
import { GET_POST_QUERY } from '../../apollo/queries'

const cache = new InMemoryCache({ addTypename: false })
const mockPost = {
  id: 42,
  title: 'Hello World',
  content: 'This is the hello world content',
  author: {
    email: 'alice@prisma.io',
  },
  published: false,
}
const mocks = [
  {
    request: {
      query: GET_POST_QUERY,
      variables: {
        id: '42',
      },
    },
    result: {
      data: { post: mockPost },
    },
  },
]

describe('Single post page', () => {
  afterEach(cleanup)

  it('renders without error', () => {
    render(<SinglePost />, {
      router: { query: { id: '42' } },
      mocks,
      cache,
      addTypename: false,
    })
  })

  it('matches snapshot', async () => {
    const { container } = render(<SinglePost />, {
      router: { query: { id: '42' } },
      mocks,
      cache,
      addTypename: false,
    })

    waitFor(() => expect(container).toMatchSnapshot())
  })
})
