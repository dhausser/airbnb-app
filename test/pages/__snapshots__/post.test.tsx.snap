// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Single post page matches snapshot 1`] = `
<div>
  <div
    class="container"
  >
    <div
      class="main"
    >
      <div
        class="grid"
      >
        <div
          class="card"
        >
          <form>
            
            <div>
              <label
                for="title-input"
                id="title-label"
              >
                Title
                 
              </label>
              <br />
              <input
                autocomplete="title"
                id="title-input"
                name="title"
                type="text"
                value="Hello World"
              />
            </div>
            <div>
              <label
                for="content-input"
                id="content-label"
              >
                Content
                 
              </label>
              <br />
              <input
                autocomplete="content"
                id="content-input"
                name="content"
                type="text"
                value="This is the hello world content"
              />
            </div>
            <button
              type="submit"
            >
              Create
            </button>
             or
             
            <a
              href="/signup"
            >
              Sign up
            </a>
          </form>
        </div>
        <button>
          Update
        </button>
        <button>
          Delete
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`Single post page matches snapshot 2`] = `<div />`;
